workspace: Add action to move focused panel to next dock position

dinocosta opened the pull request:
This Pull Request introduces a new command `workspace: move focused panel to next position` which finds the currently focused panel, if such panel exists, and moves it to the next valid dock position, following the order of `Left → Bottom → Right` and then starting again from the left position.

In order to achieve this the following changes have been introduced:

* Add a new default implementation for `PanelHandle`, namely `PanelHandle::move_to_next_position` which leverages  `PanelHandle::position`, `PanelHandle::position_is_valid` and `PanelHandle::set_position` methods to update the panel's position to the next valid position.
* Add a new method to the `workspace` module, ` move_focused_panel_to_next_position`, which is responsible for finding the currently focused panel, if such a panel exists, and calling the `move_to_next_position` method in the panel's handle.
* Add a new action to the `workspace` module, `MoveFocusedPanelToNextPosition`, which is handled by the `move_focused_panel_to_next_position` method.

Tests have also been added to the `workspace` module in order to guarantee that the action is correctly updating the focused panel's position.

Here's a quick video of it, in action 🔽 

https://github.com/user-attachments/assets/264d382b-5239-40aa-bc5e-5d569dec0734

Closes #23115 

Release Notes:

- Added new command to move the focused panel to the next valid dock position – `workspace: move focused panel to next position` .

dinocosta committed:
feat(panel): add next_position method to panel trait

* Add a new method to the `Panel` trait - `next_position` - which moves
the panel to the next valid position, following an order of
`DockPosition::Right`, `DockPosition::Bottom` and `DockPosition::Left`.
* Add a new ation to both `TerminalPanel` and `OutlinePanel` -
`NextDockPosition` - which simply calls the panel's `next_position`
default implementation.

dinocosta committed:
refactor: change method name to move_to_next_position

Tackle the feedback provided in zed-industries/zed#23155 and rename
the method from `next_position` to `move_to_next_position` while
updating both `TerminalPanel` and `OutlinePanel` to call the new method.

dinocosta committed:
refactor: move next dock position to workspace

* Remove the `NextDockPosition` action that had been introduced to both
`TerminalPanel` and `OutlinePanel`, in favour of a
`MoveFocusePanelToNextPosition` action in the workspace namespace.
* Move the `move_to_next_position` default method implementation from
the `Panel` trait to the `PanelHandle` trait, as that is the type of
the value returned by `Dock::active_panel`.
* Introduce the `Workspace::move_focused_panel_to_next_position`
method, which is triggered by the `MoveFocusedPanelToNextPosition`
action and will find thec currently focused panel in the currently
focused dock and move that panel to the next dock position.

dinocosta committed:
test: test new function to move focused panel to next dock position

Update the tests in the `workspace` to include tests for the function
that handles the `MoveFocusedPanelToNextPosition` –
`move_focused_panel_to_next_position` – making sure that everytime this
action is dispatched, a panel moves from the right to the left dock,
then to the bottom dock and, finally, to the right dock again.

SomeoneToIgnore reviewed (approved):
Thank you!

dinocosta commented:
@SomeoneToIgnore Thank you so much for reviewing and approving these changes 🙌 

Just noticed that the CI/CD pipeline for this Pull Request is failing and it seems to be linting issues regarding the code I added. I'll take a look and try to fix those ASAP!

SomeoneToIgnore commented:
Thank you for noticing.
Running `./script/clippy` on your branch would be enough to reproduce these errors.

dinocosta committed:
style: fix lint issues for option_map_unit_fn

Prefer the use of `if let Some` instead of `.map` on `Option` in the
`move_focused_panel_to_next_position` method.

dinocosta commented:
> Thank you for noticing. Running `./script/clippy` on your branch would be enough to reproduce these errors.

Thanks @SomeoneToIgnore ! I've managed to reproduce and fix these errors with https://github.com/zed-industries/zed/pull/23317/commits/9bb8f7eadce273b283237de0f9e163f22a3132a6 🧹
